hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/transfer/customfees/CustomFractionalFeeAssessor.java

```diff
@@ -0,0 +1,246 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hedera.node.app.service.token.impl.handlers.transfer.customfees;
+
+import static com.hedera.hapi.node.base.ResponseCodeEnum.CUSTOM_FEE_MUST_BE_POSITIVE;
+import static com.hedera.hapi.node.base.ResponseCodeEnum.CUSTOM_FEE_OUTSIDE_NUMERIC_RANGE;
+import static com.hedera.hapi.node.base.ResponseCodeEnum.INSUFFICIENT_SENDER_ACCOUNT_BALANCE_FOR_CUSTOM_FEE;
+import static com.hedera.node.app.service.token.impl.handlers.transfer.customfees.AdjustmentUtils.asFixedFee;
+import static com.hedera.node.app.service.token.impl.handlers.transfer.customfees.AdjustmentUtils.getFungibleTokenCredits;
+import static com.hedera.node.app.service.token.impl.handlers.transfer.customfees.AdjustmentUtils.safeFractionMultiply;
+import static com.hedera.node.app.service.token.impl.handlers.transfer.customfees.CustomFeeExemptions.isPayerExempt;
+import static com.hedera.node.app.spi.workflows.HandleException.validateTrue;
+
+import com.hedera.hapi.node.base.AccountID;
+import com.hedera.hapi.node.base.TokenID;
+import com.hedera.hapi.node.transaction.AssessedCustomFee;
+import com.hedera.hapi.node.transaction.CustomFee;
+import com.hedera.hapi.node.transaction.FractionalFee;
+import com.hedera.node.app.spi.workflows.HandleException;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.util.HashMap;
+import java.util.Map;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+
+/**
+ * Assesses fractional custom fees for given token transfer.
+ * All fractional fees, that are not netOfTransfers will manipulate the given transaction body.
+ * If netOfTransfers flag is set to false, the custom fee si reclaimed from the credits in
+ * given transaction body.
+ * If netOfTransfers flag is set to true the sender will pay the custom fees.Else the receivers will pay custom fees This means that the fee will be charged from the sender account. This is done to avoid
+ * manipulate the given transaction
+ */
+@Singleton
+public class CustomFractionalFeeAssessor {
+    private final CustomFixedFeeAssessor fixedFeeAssessor;
+
+    @Inject
+    public CustomFractionalFeeAssessor(CustomFixedFeeAssessor fixedFeeAssessor) {
+        this.fixedFeeAssessor = fixedFeeAssessor;
+    }
+
+    /**
+     * Calculates the custom fee amount that should be paid for given fraction fee.
+     * If the fee is netOfTransfers the sender will pay the fee, otherwise the receiver will pay the fee
```

Comment:
Maybe,
```
If the fee is netOfTransfers the sender will pay the fee, otherwise the receiver(s) will effectively pay the fee, as they are the accounts whose balances will be lower than if the fee had not existed.
```